File: study/lua/src/app/scene/view/petDetail/PetDetailSkillCell.lua
--
-- Author: hedili
-- Date: 2018-01-25 16:00:02
--
local ListViewCellBase = require("app.ui.ListViewCellBase")
local PetDetailSkillCell = class("PetDetailSkillCell", ListViewCellBase)
local HeroSkillActiveConfig = require("app.config.hero_skill_active")
local UserDataHelper = require("app.utils.UserDataHelper")
function PetDetailSkillCell:ctor(skillId,showSkillDetail,petBaseId, petStar, petUnitData)
	self._skillId = skillId
	self._label = nil --技能描述文本
	self._showSkillDetail = showSkillDetail
	self._petBaseId = petBaseId or 0
	self._petStar = petStar or 0 
	self._petUnitData = petUnitData
	local resource = {
		file = Path.getCSB("PetDetailSkillCell", "pet"),
		binding = {
			_btnSkill = {
				events = {{event = "touch", method = "_onButtonSkill"}}
			},
		},
	}
	PetDetailSkillCell.super.ctor(self, resource)
end

function PetDetailSkillCell:onCreate()
	if self._showSkillDetail == false then
		self._btnSkill:setVisible(false)
	end
	local contentSize = self._panelBg:getContentSize()
	local height = contentSize.height

	local config = HeroSkillActiveConfig.get(self._skillId)

	local pendingSkill = ""

	if self._petBaseId > 0 then
		local petStarCfg = UserDataHelper.getPetStarConfig(self._petBaseId, self._petStar )
		
		if petStarCfg.skill2 == self._skillId then
			pendingSkill = petStarCfg.chance_description
		end
	end

	if config then
		local skillIconRes = config.skill_icon
		local skillDes = "["..config.name.."]"..config.description
		self._imageSkillIcon:loadTexture(Path.getCommonIcon("skill", skillIconRes))
		if self._label == nil then
			self._label = cc.Label:createWithTTF("", Path.getCommonFont(), 20)
			self._label:setColor(Colors.BRIGHT_BG_TWO)
			self._label:setWidth(260)
			self._label:setAnchorPoint(cc.p(0, 1))
			self._panelBg:addChild(self._label)
		end
		self._label:setString(skillDes..pendingSkill)
		local desHeight = self._label:getContentSize().height + 35
		height = math.max(contentSize.height, desHeight)--上下各扩展5像素
		self._label:setPosition(cc.p(120, height - 5))
	else
		logError(string.format("hero_skill_active config can not find id = %d", self._skillId))
	end
	
	local size = cc.size(contentSize.width, height) 
	self:setContentSize(size)
	self._imageSkillBg:setPosition(cc.p(17, height - 1))
	self._imageBg:setContentSize(cc.size(contentSize.width, height - 2))

	self._btnSkill:setPositionY(4)
end

--点击神兽技能详情
function PetDetailSkillCell:_onButtonSkill( ... )
	-- body
	local petId = G_UserData:getPet():getCurPetId()
	local skillShowList = require("app.scene.view.petDetail.PetDetailViewHelper").makeSkillEx(petId, self._skillId, self._petUnitData)
	local PopupSkillDetail = require("app.ui.PopupSkillDetail").new(skillShowList)
	PopupSkillDetail:openWithAction()
end
return PetDetailSkillCell

File: study/ts/script/scene/view/pet/PetDetailSkillCell.ts
import ListViewCellBase from "../../../ui/ListViewCellBase";
import UIHelper from "../../../utils/UIHelper";
import { G_ConfigLoader, Colors, G_UserData } from "../../../init";
import { ConfigNameConst } from "../../../const/ConfigNameConst";
import { UserDataHelper } from "../../../utils/data/UserDataHelper";
import { Path } from "../../../utils/Path";
import { PetDetailViewHelper } from "./PetDetailViewHelper";
import PopupSkillDetail from "../../../ui/PopupSkillDetail";

const { ccclass, property } = cc._decorator;

@ccclass
export default class PetDetailSkillCell extends ListViewCellBase {

    @property({
        type: cc.Node,
        visible: true
    })
    _panelBg: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBg: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageSkillBg: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageSkillIcon: cc.Sprite = null;

    @property({
        type: cc.Button,
        visible: true
    })
    _btnSkill: cc.Button = null;

    @property(cc.Label)
    desText: cc.Label = null;

    private _skillId: number;
    private _label: cc.Label;
    private _showSkillDetail: boolean;
    private _petBaseId: number;
    private _petStar: number;
    private _petUnitData: any;
    private _isLoaded: boolean;

    ctor(skillId, showSkillDetail, petBaseId, petStar, petUnitData) {
        this._skillId = skillId;
        this._showSkillDetail = showSkillDetail;
        this._petBaseId = petBaseId || 0;
        this._petStar = petStar || 0;
        this._petUnitData = petUnitData;
        UIHelper.addEventListener(this.node, this._btnSkill, 'PetDetailSkillCell', '_onButtonSkill');
    }
    onCreate() {
        if (this._isLoaded) {
            return;
        }
        this._isLoaded = true;
        if (this._showSkillDetail == false) {
            this._btnSkill.node.active = (false);
        }
        var contentSize = this._panelBg.getContentSize();
        var height = contentSize.height;
        var config = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_ACTIVE).get(this._skillId);
        var pendingSkill = '';
        if (this._petBaseId > 0) {
            var petStarCfg = UserDataHelper.getPetStarConfig(this._petBaseId, this._petStar);
            if (petStarCfg.skill2 == this._skillId) {
                pendingSkill = petStarCfg.chance_description;
            }
        }
        if (config) {
            var skillIconRes = config.skill_icon;
            var skillDes = '[' + (config.name + (']' + config.description));
            UIHelper.loadTexture(this._imageSkillIcon, Path.getCommonIcon('skill', skillIconRes));
            if (this._label == null) {
                // this._label = UIHelper.createWithTTF('', Path.getCommonFont(), 20);
                this._label = this.desText;
                this._label.node.color = (Colors.BRIGHT_BG_TWO);
                //this._label.node.width = (260);
                this._label.node.setAnchorPoint(cc.v2(0, 1));
                //this._panelBg.addChild(this._label.node);
            }
            this._label.string = (skillDes + pendingSkill);
            UIHelper.updateLabelSize(this._label);
            var desHeight = this._label.node.getContentSize().height + 35;
            height = Math.max(contentSize.height, desHeight);
            this._label.node.setPosition(cc.v2(120, height - 5));
        } else {
            //logError(string.format('hero_skill_active config can not find id = %d', this._skillId));
        }
        var size = cc.size(contentSize.width, height);
        this.node.setContentSize(size);
        this._imageSkillBg.node.setPosition(cc.v2(17, height - 1));
        this._imageBg.node.setContentSize(cc.size(contentSize.width, height - 2));
        this._btnSkill.node.y = (4);
    }
    _onButtonSkill() {
        var petId = G_UserData.getPet().getCurPetId();
        var skillShowList = PetDetailViewHelper.makeSkillEx(petId, this._skillId, this._petUnitData);
        PopupSkillDetail.getIns(PopupSkillDetail, (p: PopupSkillDetail) => {
            p.ctor(skillShowList);
            p.openWithAction();
        })
    }
}
